rootfs/etc/nginx/lua/util/dns.lua

```diff
@@ -28,48 +27,65 @@ local function a_records_and_max_ttl(answers)
   return addresses, ttl
 end
 
-function _M.resolve(host)
-  local cached_addresses = cache:get(host)
-  if cached_addresses then
-    local message = string.format(
-      "addresses %s for host %s was resolved from cache",
-      table.concat(cached_addresses, ", "), host)
-    ngx.log(ngx.INFO, message)
-    return cached_addresses
-  end
-
+local function resolve_host(host, qtype)
   local r
   r, err = resolver:new{
-    nameservers = util.deepcopy(configuration.nameservers),
+    nameservers = configuration.nameservers,
     retrans = 5,
     timeout = 2000,  -- 2 sec
   }
 
   if not r then
-    ngx.log(ngx.ERR, "failed to instantiate the resolver: " .. tostring(err))
-    return { host }
+    return nil, -1, "failed to instantiate the resolver: " .. tostring(err)
   end
 
   local answers
-  answers, err = r:query(host, { qtype = r.TYPE_A }, {})
+  answers, err = r:query(host, { qtype = qtype }, {})
   if not answers then
-    ngx.log(ngx.ERR, "failed to query the DNS server: " .. tostring(err))
-    return { host }
+    return nil, -1, "failed to query the DNS server: " .. tostring(err)
   end
 
   if answers.errcode then
-    ngx.log(ngx.ERR, string.format("server returned error code: %s: %s", answers.errcode, answers.errstr))
-    return { host }
+    return nil, -1, string.format("server returned error code: %s: %s", answers.errcode, answers.errstr)
   end
 
   local addresses, ttl = a_records_and_max_ttl(answers)
   if #addresses == 0 then
-    ngx.log(ngx.ERR, "no A record resolved")
+    return nil, -1, "no record resolved"
+  end
+
+  return addresses, ttl, nil
+end
+
+function _M.resolve(host)
+  local cached_addresses = cache:get(host)
+  if cached_addresses then
+    local message = string.format(
+      "addresses %s for host %s was resolved from cache",
+      table.concat(cached_addresses, ", "), host)
+    ngx.log(ngx.INFO, message)
+    return cached_addresses
+  end
+
+  local addresses, ttl
+  addresses, ttl, err = resolve_host(host, resolver.TYPE_A)
+  if not addresses then
+    ngx.log(ngx.ERR, "failed to query the DNS server: " .. tostring(err))
```

Comment:
I'd not log an error here, because it might be expected that there's no `TYPE_A` record, so we will keep logging error.

Instead I suggest you collect this and the error below and log them in the end only if both TYPE_A and TYPE_AAAA queries fail